src/library_wasi.js

```diff
@@ -263,24 +263,159 @@ var WasiLibrary = {
+                      fdflags, opened_fd) {
+    if (!(fd in preopens)) {
+      return {{{ cDefine('EBADF') }}};
+    }
+    var pathname = UTF8ToString(path, path_len);
+    var musl_oflags = wasiRightsToMuslOFlags(Number(fs_rights_base));
+#ifdef SYSCALL_DEBUG
+    err("oflags1: 0x" + musl_oflags.toString(16));
+#endif
+    musl_oflags |= wasiOFlagsToMuslOFlags(Number(oflags));
+#ifdef SYSCALL_DEBUG
+    err("oflags2: 0x" + musl_oflags.toString(16));
+#endif
+    var stream = FS.open(pathname, musl_oflags);
+    {{{ makeSetValue('opened_fd', '0', 'stream.fd', 'i32') }}};
+    return 0;
+  },
+
+  fd_prestat_dir_name__deps: ['$preopens'],
+  fd_prestat_dir_name__sig: 'iiii',
+  fd_prestat_dir_name__nothrow: true,
+  fd_prestat_dir_name: function(fd, path, path_len) {
+    if (!(fd in preopens)) {
+      return {{{ cDefine('EBADF') }}};
+    }
+    var preopen = preopens[fd];
+    stringToUTF8Array(preopens, HEAP8, path, path_len)
+    return 0;
+  },
+
+  fd_prestat_get__deps: ['$preopens'],
+  fd_prestat_get__sig: 'iii',
+  fd_prestat_get__nothrow: true,
+  fd_prestat_get: function(fd, stat_buf) {
+    if (!(fd in preopens)) {
+      return {{{ cDefine('EBADF') }}};
+    }
+    var preopen = preopens[fd];
+    {{{ makeSetValue('stat_buf', C_STRUCTS.__wasi_prestat_t.pr_type, cDefine('__WASI_PREOPENTYPE_DIR'), 'i8') }}};
+    {{{ makeSetValue('stat_buf', C_STRUCTS.__wasi_prestat_t.u + C_STRUCTS.__wasi_prestat_dir_t.pr_name_len, 'preopen.length', 'i64') }}};
+    return 0;
+  },
+
+  fd_fdstat_set_flags__sig: 'iii',
+  fd_fdstat_set_flags: function(fd, flags) {
+    var stream = SYSCALLS.getStreamFromFD(fd);
+    return 0;
+  },
+
+  fd_filestat_get__sig: 'iii',
+  fd_filestat_get: function(fd, stat_buf) {
+    var stream = SYSCALLS.getStreamFromFD(fd);
+    {{{ makeSetValue('stat_buf', C_STRUCTS.__wasi_filestat_t.dev, '0', 'i64') }}};
+    {{{ makeSetValue('stat_buf', C_STRUCTS.__wasi_filestat_t.ino, '0', 'i64') }}};
+    {{{ makeSetValue('stat_buf', C_STRUCTS.__wasi_filestat_t.filetype, '0', 'i8') }}};
+    {{{ makeSetValue('stat_buf', C_STRUCTS.__wasi_filestat_t.nlink, '0', 'i64') }}};
+    {{{ makeSetValue('stat_buf', C_STRUCTS.__wasi_filestat_t.size, '0', 'i64') }}};
+    {{{ makeSetValue('stat_buf', C_STRUCTS.__wasi_filestat_t.atim, '0', 'i64') }}};
+    {{{ makeSetValue('stat_buf', C_STRUCTS.__wasi_filestat_t.mtim, '0', 'i64') }}};
+    {{{ makeSetValue('stat_buf', C_STRUCTS.__wasi_filestat_t.ctim, '0', 'i64') }}};
+    return 0;
+  },
+
+  fd_pread__sig: 'iiiiii',
+  fd_pread: function(fd, iov, iovcnt, offset, pnum) {
+    var stream = SYSCALLS.getStreamFromFD(fd)
+    var offset = SYSCALLS.get64(low, high);
+    var num = SYSCALLS.doReadv(stream, iov, iovcnt, offset);
+    {{{ makeSetValue('pnum', 0, 'num', 'i32') }}}
+    return 0;
+  },
+#endif
+
+#ifdef PURE_WASI
+  fd_fdstat_get__deps: ['$preopens'],
+#endif
   fd_fdstat_get__sig: 'iii',
   fd_fdstat_get: function(fd, pbuf) {
+    var rights_base = 0;
+    var rights_inheriting = 0;
```

Comment:
technically these should be `rightsBase` etc. in `camelCase`. But maybe it makes sense to have an exception here?